user_groups: Allow adding a user to groups via user profile.

shuklamaneesh23 opened the pull request:
This pull request introduces several enhancements to the user profile modal. The changes primarily focus on adding functionality to manage user group memberships, including the ability to add users to groups and filter the group list.

[CZO Discussion](https://chat.zulip.org/#narrow/channel/101-design/topic/Adding.20a.20user.20to.20groups.20via.20user.20profile)

Fixes: #32488.

<!-- Describe your pull request here.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>User with no User groups joined:</summary>
<img width="523" alt="Screenshot 2025-01-07 at 8 46 55 PM" src="https://github.com/user-attachments/assets/c991c3cf-57f3-494f-a8f3-9947b4b49362" />

</details>

<details>
<summary>User with a joined user group:</summary>
<img width="526" alt="Screenshot 2025-01-07 at 8 47 04 PM" src="https://github.com/user-attachments/assets/383ab094-6c74-48dd-b1b9-c1261c736522" />

</details>

<details>
<summary>Typeahead showing user groups:</summary>
<img width="527" alt="Screenshot 2025-01-08 at 8 55 16 PM" src="https://github.com/user-attachments/assets/08555215-80f6-4298-883b-c0baade6e54e" />

</details>


<details>
<summary>User added successfully:</summary>
<img width="524" alt="Screenshot 2025-01-08 at 8 55 36 PM" src="https://github.com/user-attachments/assets/1b0de6e6-862d-4faa-bb24-43c60178c514" />

</details>

<details>
<summary>Filter with no match:</summary>
<img width="521" alt="Screenshot 2025-01-07 at 8 50 08 PM" src="https://github.com/user-attachments/assets/0583f985-1128-4e0d-a7cf-25cbd2db7fff" />
</details>

<details>
<summary>User with no permission:</summary>
<img width="517" alt="Screenshot 2025-01-08 at 8 56 29 PM" src="https://github.com/user-attachments/assets/ea17e249-2641-4951-b38f-c073a62faff3" />

</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

shuklamaneesh23 commented:
Hey @alya,  
Please take a look at the UI changes. If they look good to you, kindly assign the it to some maintainer, as I’m planning to refactor some code to reduce duplication due to the similarities between the `stream subscription` section and the `user group` UI. 

Thanks!

alya commented:
My note about errors was that we need to make sure something reasonable happens if you add a user to 5 groups and there's an error on the 3rd one. I expect you can find examples of this sort of thing elsewhere in the app.

alya commented:
If a user doesn't have permission to add someone to groups, we don't need a banner. We should just not show the UI for adding users.

alya commented:
The confirmation banner is styled differently here from the one for adding users to channels. You can post screenshots of both in #design to see what folks prefer.

shuklamaneesh23 commented:
Hi @alya, 

I have addressed all the reviews except for the last one, as I didn’t fully understand what you meant there. However, I’ve posted screenshots of both the success banners in the [CZO](https://chat.zulip.org/#narrow/channel/101-design/topic/Adding.20a.20user.20to.20groups.20via.20user.20profile).

Please let me know if any further modifications are required.  
Thanks!

alya commented:
I was confused about the banners. All set.

alya commented:
Works for me in manual testing! I haven't tested permissions configurations.

alya commented:
@sahil839 are you up for reviewing this release priority PR?

sahil839 commented:
@shuklamaneesh23 Can you move the changes to add search input in a separate commit?

shuklamaneesh23 commented:
Hi @sahil839 , I have done so.

sahil839 commented on web/templates/user_profile_modal.hbs:
```diff
@@ -131,8 +131,35 @@
+                        </div>
                         <div class="subscription-group-list empty-list">
-                            <table class="user-group-list" data-empty="{{t 'No user group subscriptions.'}}"></table>
+                            <table class="user-group-list" data-empty="{{t 'Not a member of any groups.'}}" data-search-results-empty="{{t 'No matching user groups' }}"></table>
```
This should be in the commit which adds the search input and probably a separate commit to improve the `data-empty` value.

sahil839 commented on web/src/user_profile.ts:
```diff
@@ -80,9 +82,12 @@ export type CustomProfileFieldData = {
 let toggler: components.Toggle;
 let bot_owner_dropdown_widget: DropdownWidget | undefined;
 let original_values: (Record<string, unknown> & {user_id?: string | undefined}) | undefined;
+let show_user_group_container =
```
Why is this required here? I see it is being used in the "Add" button click handler, but user cannot click on the button if this is not visible.

shuklamaneesh23 replied:
This determines whether a user has access to add someone to a user group or not.  
I have used it inside the `show_user_profile` function, but as you mentioned, I also need it in the `initialize` function under the `Add` button controller. If I don't define it outside the scope of the `show_user_profile` function, I won't be able to access it there.

sahil839 replied:
I am saying why it is needed in `Add` button click handler. Button will only be visible if `show_user_group_container` will be true.

shuklamaneesh23 replied:
Yes, as Alya specified [here](https://github.com/zulip/zulip/pull/32941#issuecomment-2575817650) that if a user have not permission to add a user to a user group, then we don't want to show that pill-container and add button.

sahil839 replied:
I think you missed to remove this.

sahil839 commented on web/src/user_profile.ts:
```diff
@@ -546,6 +551,8 @@ export function show_user_profile(user: User, default_tab_key = "profile-tab"):
         !user.is_system_bot &&
         people.is_person_active(user.user_id);
+    show_user_group_container =
+        invite_user_group_picker_pill.get_user_groups_allowed_to_add_members().length > 0;
```
Might be good to move `get_user_groups_allowed_to_add_members()` function to `user_groups` in a prep commit.

sahil839 replied:
We have all the data functions like `get_user_groups_allowed_to_mention`, `get_potential_subgroups` etc. in `user_groups.ts`, so that might be a better place.

sahil839 commented on web/templates/user_profile_modal.hbs:
```diff
@@ -131,8 +131,35 @@
+                        {{/if}}
+                        <div class="stream-list-top-section">
+                            <div class="header-section">
+                                <h3 class="stream-tab-element-header">{{t 'Group membership' }}</h3>
```
Would be good to add this heading as part of the commit which adds the search input and probably make that the first commit of this PR. Keeping this commit only for adding the widget to add user to groups.

sahil839 commented on web/src/add_subscribers_pill.ts:
```diff
@@ -101,19 +101,22 @@ export function generate_pill_html(item: CombinedPill): string {
-    const $pill_widget_button = $pill_container.parent().find(".add-users-button");
-
+    const $pill_widget_button = $pill_container.parent().hasClass("add-user-group-container")
+        ? $pill_container.parent().parent().parent().find(".add-groups-button")
```
Would be better to just add a common class for all the add buttons and use that directly here.

sahil839 commented on web/src/add_subscribers_pill.ts:
```diff
@@ -101,19 +101,22 @@ export function generate_pill_html(item: CombinedPill): string {
         $pill_widget_button.prop("disabled", pill_widget.items().length === 0),
     );
+    // If a pill is added, enable the add button.
+    pill_widget.onPillCreate(() => $pill_widget_button.prop("disabled", false));
```
Why is this needed now but not before? If this fixes some pre-existing bug, this change should be in a separate commit.

shuklamaneesh23 replied:
This function was previously used in the input pill container for adding members to a user group while creating that user group.

In that case, the typeahead appeared only after typing a few letters into the input box. However, in this scenario, the typeahead appears as soon as the user places the cursor inside the input box, allowing them to select a pill without typing anything. This extended code accommodates that behavior.

I am attaching a demo video of both cases to clarify the need for this change.

<details>
<summary> User group setting section: (typeahead after typing some words) </summary>

https://github.com/user-attachments/assets/73b48d83-8799-4668-9444-a876bceb6d4c


</details>

<details>
<summary> Current area of use:(typeahead without typing a single word) </summary>

https://github.com/user-attachments/assets/5f0d9c41-cd3a-457b-b420-32cceb8b77c0


</details>

shuklamaneesh23 replied:
No, it is not discussed on CZO. However the design that Alya posted in the issue description was of the `user-group-pill-container` of invite modal from where I took this reference.

sahil839 commented on web/src/invite_user_group_picker_pill.ts:
```diff
@@ -62,7 +85,10 @@ function generate_pill_html(item: UserGroupPill): string {
 }
 
-export function create($user_group_pill_container: JQuery): user_group_pill.UserGroupPillWidget {
+export function create(
```
It makes sense to use the functions in this file to avoid duplication, but then we might need to rename this file since this functions are not just used for invite.

shuklamaneesh23 replied:
Yeah, I think so, if you allow I am planning to rename it with `user_group_picker_pill.ts`.

shuklamaneesh23 replied:
Sure!

sahil839 commented on web/src/invite_user_group_picker_pill.ts:
```diff
@@ -43,9 +51,24 @@ export function get_user_groups_allowed_to_add_members(): UserGroup[] {
-function set_up_pill_typeahead({pill_widget, $pill_container}: SetUpPillTypeaheadConfig): void {
+export function get_user_groups_where_user_can_be_added(user_id: number): UserGroup[] {
+    const all_user_groups: UserGroup[] = user_groups.get_realm_user_groups();
+    user_groups_of_user = user_groups.get_user_groups_of_user(user_id);
```
This will also include groups which the user is not a direct member of but is member of any of its subgroup and thus will not be shown in the typeahead. But we want to show them in the typeahead, because I guess we would still want to allow user to be added to such groups.

sahil839 commented on web/src/invite_user_group_picker_pill.ts:
```diff
@@ -43,9 +51,24 @@ export function get_user_groups_allowed_to_add_members(): UserGroup[] {
     const user_group_source: () => UserGroup[] = () => {
-        const groups_with_permission = get_user_groups_allowed_to_add_members();
+        const groups_with_permission = user_id
+            ? get_user_groups_where_user_can_be_added(user_id)
```
This function does not check permission to add members to the groups. A better way would be something like -
```js
let groups_allowed = get_user_groups_allowed_to_add_members();
if (user_id !== undefined) {
    groups_allowed = groups_allowed.filter((group) => // logic to include only the groups which user is not a member of)
}
```

shuklamaneesh23 replied:
Instead of performing the filtering here, I am doing it in the function `get_user_groups_where_user_can_be_added` and conditionally calling it here, making the code more readable.

shuklamaneesh23 replied:
Thanks for this `Sahil`. It is working well and also reducing the lines of code.😀

sahil839 commented on web/templates/user_profile_modal.hbs:
```diff
@@ -157,6 +157,12 @@
                             <div class="header-section">
                                 <h3 class="stream-tab-element-header">{{t 'Group membership' }}</h3>
                             </div>
+                            <div class="stream-search-container">
```
Can we replace `stream` with `group` in the classes used for this block?

shuklamaneesh23 replied:
If we do so, we need to add these class names to the css file at all the places where stream is used.
Like if previously it is :
```
.stream-search-container {
             // some css
          }
```

Then, I need to use:
```
.groups-search-container,
.stream-search-container {
             // some css
          }
```

sahil839 reviewed (commented):
Posted some comments.

shuklamaneesh23 commented:
Hi @sahil839, 

First of all, thank you for such a dedicated review.  
I have made the changes based on your comments wherever possible and have also responded to the questions you raised.  

If anything remains unclear from my messages above, please let me know.  

Thanks!

sahil839 commented on web/src/invite_user_group_picker_pill.ts:
```diff
@@ -31,6 +34,11 @@ function create_item_from_group_name(
 
+    // check if the user is already a member of the user group.
+    if (user_groups_of_user.some((user_group: UserGroup) => user_group.id === group.id)) {
+        return undefined;
```
It would be good if we just direct use `is_direct_member_of` check here instead of using a module-level variable to store the value.

shuklamaneesh23 replied:
Doing so would require the access of `user_id` inside this function and need a lot more function definitions change, including the main parent file `input_pill` which I don't think going to worth us.

shuklamaneesh23 commented:
Hi @**sahil839**, I have updated the code as per the CZO discussion.  

Now, the addition of a user to groups happens in series, and the process stops adding the user to further groups upon encountering an error. It displays an error message indicating the group where the error occurred and leaves the remaining group pills in the pill container as they are.  

Please have another look.  
Thanks for the review and the guidance to achieve this behaviour.

sahil839 commented on web/src/user_profile.ts:
```diff
@@ -540,6 +545,67 @@ export function show_user_profile_access_error_modal(): void {
+                );
+                if ($pill.length > 0) {
+                    $pill.find(".exit").first().trigger("click");
+                }
```
Instead of removing these pills one by one, would be better to just clear it once - either when user has been added to all groups successfully or after we get the first error.

sahil839 commented on web/src/user_profile.ts:
```diff
@@ -540,6 +545,67 @@ export function show_user_profile_access_error_modal(): void {
 }
 
+function add_user_to_groups(group_ids: number[], user_id: number, $alert_box: JQuery): void {
+    let success_count = 0;
```
Might be good to maintain the list of groups IDs here instead of this count in `group_ids_successfully_added` as that might be useful to remove the pills on receiving first error.

shuklamaneesh23 replied:
As we are going with Tim's suggestion of making series additions, he suggested [here](https://chat.zulip.org/#narrow/channel/101-design/topic/Adding.20a.20user.20to.20groups.20via.20user.20profile/near/2036957) to not maintain an array.
Just let me know if I'm getting it wrong.

sahil839 commented on web/src/user_profile.ts:
```diff
@@ -540,6 +545,67 @@ export function show_user_profile_access_error_modal(): void {
+    let success_count = 0;
+
+    // Recursive function to process groups in series (one after another)
+    function add_user_to_next_group(index: number): void {
```
Would be better for this function to just accept group ID as parameter. And then to find next group ID can use length of `group_ids_successfully_added` to find index for next group ID.

sahil839 commented on web/src/input_pill.ts:
```diff
@@ -135,7 +137,12 @@ export function create<ItemType extends {type: string}>(
+            const item = store.create_item_from_text(
+                text,
+                existing_items,
+                store.pill_config,
```
I meant including `user_id` as part of `pill_config`.

sahil839 commented on web/src/user_profile.ts:
```diff
@@ -540,6 +543,68 @@ export function show_user_profile_access_error_modal(): void {
+                          {group_name: target_user_group.name},
+                      );
+
+                ui_report.client_error(error_message, $alert_box, 1200);
```
I think the error message is cleared too soon. Maybe we can also just keep it and not hide it but am not sure.

shuklamaneesh23 replied:
I will post a question regarding this in the CZO.

shuklamaneesh23 replied:
Hi @timabbott , sorry for that.
Here, is the [CZO link](https://chat.zulip.org/#narrow/channel/101-design/topic/Adding.20a.20user.20to.20groups.20via.20user.20profile/near/2043212) for this question.

sahil839 reviewed (commented):
Posted a couple of comments.

I noticed one bug that if I add or remove a user from some groups from user profile modal and then go to "Channels" tab and then "Groups" tab again the groups list is not correct until I refresh. But this might not be related to this PR and bug could be there in main as well.

shuklamaneesh23 commented:
> I noticed one bug that if I add or remove a user from some groups from user profile modal and then go to "Channels" tab and then "Groups" tab again the groups list is not correct until I refresh. But this might not be related to this PR and bug could be there in main as well.

Yes @sahil839 , this bug is present on main (for channels as well as for groups) and it could be fixed separately. Filed an issue regarding this on the [CZO](https://chat.zulip.org/#narrow/channel/9-issues/topic/groups.2Fchannels.20list.20is.20not.20correct.20until.20I.20refresh).


https://github.com/user-attachments/assets/37158510-c197-4313-85ab-14857acf6e0a


https://github.com/user-attachments/assets/d6911b63-072f-4f52-9304-49eeab34b2d7

shuklamaneesh23 commented:
Hi @sahil839, I have updated the code as per your comments. Along with that, I have made changes to the way we are displaying the error string as per the discussion over on [CZO](https://chat.zulip.org/#narrow/channel/101-design/topic/Adding.20a.20user.20to.20groups.20via.20user.20profile/near/2037576).

However, I haven't made any changes to your [this comment](https://github.com/zulip/zulip/pull/32941#discussion_r1916799585) as Alya mentioned [here](https://chat.zulip.org/#narrow/channel/101-design/topic/Adding.20a.20user.20to.20groups.20via.20user.20profile/near/2043269) to maintain this behavior similar to the channel subscription part.

Thanks for the review. Do share how I could further improve my code.

sahil839 commented on web/src/user_profile.ts:
```diff
@@ -540,6 +543,82 @@ export function show_user_profile_access_error_modal(): void {
+    function clear_successful_pills(): void {
+        for (const id of group_ids_successfully_added) {
+            const $pill = $(`#user-group-to-add .pill-container .pill[data-user-group-id="${id}"]`);
+            if ($pill.length > 0) {
```
Can we add a function `removePills` which accepts pill elements as an argument and just call that here?

shuklamaneesh23 replied:
@sahil839 , I have added the function as you said.
Please have another look.
Thanks!

sahil839 replied:
Sorry, I was not clear before, but I meant adding a function like `removePill` present in `input_pill.ts`.

shuklamaneesh23 replied:
@sahil839 , please have another look.
Thanks!

sahil839 reviewed (commented):
Posted one comment, looks good otherwise.

shuklamaneesh23 commented on web/src/user_profile.ts:
```diff
@@ -540,6 +543,83 @@ export function show_user_profile_access_error_modal(): void {
+            const $pill = $(`#user-group-to-add .pill-container .pill[data-user-group-id="${id}"]`);
+            const pillElement = $pill[0];
+            if ($pill.length > 0 && pillElement) {
+                user_group_pill_widget.removePill(pillElement);
```
No, I initially tried it that way, but the linter showed an error: `Argument of type 'HTMLElement | undefined' is not assignable to parameter of type 'HTMLElement'.` 

If you agree, I can pass `$pill[0]!`, which ensures that it will never be undefined."

sahil839 replied:
Ok. I am not sure about how "trigger" is used, but it is better to use "close" as trigger here, since we use "clear" when removing all pills.

sahil839 commented:
Looks good to me. @alya do you want to test this again?

alya commented:
Looks good to me in manual testing! I wasn't sure how to test the error banners, though.

alya commented:
@timabbott This PR has been approved by @sahil839 .

timabbott commented on web/src/user_profile.ts:
```diff
@@ -548,6 +551,82 @@ export function show_user_profile_access_error_modal(): void {
+            if (group_ids_successfully_added.length > 0) {
+                ui_report.success(
+                    $t_html({
+                        defaultMessage: "Successfully added to group(s)!",
```
"Successfully updated membership" might be a bit nicer for translators.

timabbott replied:
This is better, since we're looking at a user, not a group:
``` diff
diff --git a/web/src/user_profile.ts b/web/src/user_profile.ts
index e2b7277d2b..24b7440af8 100644
--- a/web/src/user_profile.ts
+++ b/web/src/user_profile.ts
@@ -559,7 +559,7 @@ function add_user_to_groups(group_ids: number[], user_id: number, $alert_box: JQ
             if (group_ids_successfully_added.length > 0) {
                 ui_report.success(
                     $t_html({
-                        defaultMessage: "Successfully added to group(s)!",
+                        defaultMessage: "Added successfully!",
                     }),
                     $alert_box,
                     1200,
```

timabbott commented:
Fixed the one thing noted above and merged, thanks for all the work on this @shuklamaneesh23 and @sahil839!

I think we will want to do a follow-up reworking of the success/error banner UX for both this and the subscribers UI, since it's not at the level of polish of the rest of the app. I'll let @alya drive the conversation around that. Might not be a 10.0 goal.

alya commented:
Yeah, I think that should be part of the overall modal redesign project @terpimost has been working on (definitely not 10.0).
